File: java/sbt/dse/src/main/java/com/datastax/spark/example/WriteRead.java
package com.datastax.spark.example;

import com.datastax.driver.core.Session;
import com.datastax.spark.connector.cql.CassandraConnector;
import com.datastax.spark.connector.japi.rdd.CassandraTableScanJavaRDD;
import com.google.common.collect.ImmutableMap;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.function.MapFunction;
import org.apache.spark.rdd.RDD;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import scala.Tuple2;
import scala.runtime.AbstractFunction1;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static com.datastax.spark.connector.japi.CassandraJavaUtil.*;


// For DSE it is not necessary to set connection parameters for spark.master (since it will be done
// automatically)
public class WriteRead {
  public static void main(String[] args) {

    // A SparkSession
    SparkSession spark = SparkSession
      .builder()
      .appName("Datastax Java example")
      .getOrCreate();

    CassandraConnector.apply(spark.sparkContext()).withSessionDo(
      new AbstractFunction1<Session, Object>() {
        public Object apply(Session session) {
          session.execute("CREATE KEYSPACE IF NOT EXISTS ks WITH "
            + "replication = {'class': 'SimpleStrategy', 'replication_factor': 1 }");
          return session
            .execute("CREATE TABLE IF NOT EXISTS ks.kv (k int, v int, PRIMARY KEY (k))");
        }
      });

    JavaRDD<Tuple2<Integer, Integer>> data = spark
      .range(1, 10)
      .javaRDD()
      .map(x -> new Tuple2<>(x.intValue(), x.intValue()));

    javaFunctions(data)
      .writerBuilder("ks", "kv", mapTupleToRow(Integer.class, Integer.class))
      .saveToCassandra();

    // Read data as RDD
    JavaRDD<Tuple2<Integer, Integer>> scReadData = javaFunctions(spark.sparkContext())
      .cassandraTable("ks", "kv", mapRowToTuple(Integer.class, Integer.class));

    // Read data as DataSet (DataFrame)
    Dataset<Row> dataset = spark
      .read()
      .format("org.apache.spark.sql.cassandra")
      .options(ImmutableMap.of("table", "kv", "keyspace", "ks"))
      .load();

    System.out.println("Data read as RDD");
    scReadData.collect()
      .forEach(System.out::println);

    System.out.println("Data read as DataSet (DataFrame)");
    dataset
      .javaRDD()
      .map(row -> new Tuple2<>(row.getInt(0), row.getInt(1)))
      .collect()
      .forEach(System.out::println);

    spark.stop();
    System.exit(0);
  }
}

File: scala/sbt/dse/src/main/scala/com/datastax/spark/example/WriteRead.scala
package com.datastax.spark.example

import com.datastax.spark.connector._
import com.datastax.spark.connector.cql.CassandraConnector
import org.apache.spark.sql.{SaveMode, SparkSession}
import org.apache.spark.sql.cassandra._

// For DSE it is not necessary to set connection parameters for spark.master (since it will be done
// automatically)
object WriteRead extends App {

  val spark = SparkSession.builder
    .appName("Datastax Scala example")
    .enableHiveSupport()
    .getOrCreate()

  import spark.implicits._

  // Create keyspace and table
  CassandraConnector(spark.sparkContext).withSessionDo { session =>
    session.execute(
      """CREATE KEYSPACE IF NOT EXISTS ks WITH
        | replication = {'class': 'SimpleStrategy', 'replication_factor': 1 }""".stripMargin)
    session.execute("""CREATE TABLE IF NOT EXISTS ks.kv (k int, v int, PRIMARY KEY (k))""")
  }

  // Write some data
  spark.range(1, 10)
    .map(x => (x, x))
    .rdd
    .saveToCassandra("ks", "kv")

  // Read data as RDD
  val rdd = spark.sparkContext
    .cassandraTable(keyspace = "ks", table = "kv")

  // Read data as DataSet (DataFrame)
  val dataset = spark.read
    .cassandraFormat(keyspace = "ks", table = "kv")
    .load()

  println("Data read as RDD")
  rdd.collect()
    .foreach(println)

  println("Data read as DataSet (DataFrame)")
  dataset.collect()
    .foreach(println)

  spark.stop()
  sys.exit(0)
}